File: frontend/src/lib/api/facilitator.ts
/** Facilitator instance-admin API calls (#1020, #1022). */

import { apiClient } from "@/lib/api-client";
import type { ApiRequestBody, ApiResponse } from "@/types/api-helpers";

export const facilitatorApi = {
    /**
     * This instance's join-link settings — lazily generates the join token on
     * first read.
     */
    getAccess: () =>
        apiClient.get<ApiResponse<"/api/v1/facilitator/access", "get">>(
            "/facilitator/access",
        ),

    /** Flip whether the join link currently admits new accounts. */
    updateAccess: (
        data: ApiRequestBody<"/api/v1/facilitator/access", "patch">,
    ) => apiClient.patch<void>("/facilitator/access", data),

    /** This instance's roster, split into joined vs invited-not-yet-joined. */
    getRoster: () =>
        apiClient.get<ApiResponse<"/api/v1/facilitator/roster", "get">>(
            "/facilitator/roster",
        ),

    /**
     * Existing accounts whose username starts with `prefix` — the add control's
     * lookup.
     */
    searchRosterCandidates: (prefix: string) =>
        apiClient.get<
            ApiResponse<"/api/v1/facilitator/roster/candidates", "get">
        >("/facilitator/roster/candidates", { params: { prefix } }),

    /** Add an existing account to the roster. */
    addToRoster: (data: ApiRequestBody<"/api/v1/facilitator/roster", "post">) =>
        apiClient.post<void>("/facilitator/roster", data),

    /** Ban/remove a username from the roster. */
    removeFromRoster: (username: string) =>
        apiClient.delete<void>(
            `/facilitator/roster/${encodeURIComponent(username)}`,
        ),
};

File: src/energetica/schemas/facilitator.py
"""Schemas for the facilitator instance-admin surface (#1020)."""

from __future__ import annotations

from pydantic import BaseModel, Field


class FacilitatorAccessOut(BaseModel):
    """This instance's join-link settings, as shown on the facilitator page."""

    join_token: str = Field(description="The unguessable token encoded in the join link; generated on first read.")
    join_open: bool = Field(description="Whether the join link currently admits new accounts.")


class FacilitatorAccessPatch(BaseModel):
    """Request body to flip whether the join link currently admits new accounts."""

    join_open: bool


class FacilitatorRosterOut(BaseModel):
    """The private instance's roster (#1022), split by whether the account has settled yet."""

    joined: list[str] = Field(description="Roster usernames that have settled (have a Player) on this instance.")
    invited: list[str] = Field(description="Roster usernames that have not settled here yet.")


class RosterCandidatesOut(BaseModel):
    """Existing accounts matching a roster-search prefix — the add control's lookup."""

    usernames: list[str] = Field(description="Existing accounts' usernames starting with the search prefix.")


class RosterAddIn(BaseModel):
    """Request body to add an existing account to the roster."""

    username: str = Field(description="An existing account's username to add to the roster.")
